xtask/src/main.rs

```diff
@@ -8,69 +8,106 @@ mod zombienet;
-    // Wait for the sovereign rollup to be ready
-    std::thread::sleep(std::time::Duration::from_secs(20));
+    //// TODO: https://github.com/thrumdev/blobs/issues/226
+    //// Wait for the sovereign rollup to be ready
+    tokio::time::sleep(std::time::Duration::from_secs(30)).await;
 
-    sovereign.test_sovereign_rollup()?;
+    sovereign.test_sovereign_rollup().await?;
 
     Ok(())
 }
 
-fn zombienet(params: crate::cli::zombienet::Params) -> anyhow::Result<()> {
+async fn zombienet(params: crate::cli::zombienet::Params) -> anyhow::Result<()> {
     let project_path = obtain_project_path()?;
-    build::build(&project_path, params.build)?;
-    let _zombienet = zombienet::Zombienet::try_new(&project_path, params.zombienet)?;
-    wait_interrupt();
+    build::build(&project_path, params.build).await?;
+    // Create the ctrl_c handle and the channel used to communicate the pgid that needs to be killed
+    let (ctrl_c_handle, tx) = ctrl_c_handler();
+    let _zombienet = zombienet::Zombienet::try_new(&project_path, params.zombienet, tx).await?;
+    ctrl_c_handle.await?;
     Ok(())
 }
 
-fn obtain_project_path() -> anyhow::Result<PathBuf> {
-    #[rustfmt::skip]
-    let project_path = duct::cmd!(
-        "sh", "-c",
-        "cargo metadata --format-version 1 | jq -r '.workspace_root'"
-    )
-    .stdout_capture()
-    .run()?;
-    Ok(PathBuf::from(str::from_utf8(&project_path.stdout)?.trim()))
+// Spawn a new task to wait for the ctrl_c event (SIGINT signal).
+// Upon receipt, it will kill any processes that would not normally be terminated when handling SIGINT.
+//
+// A channel is used to send group IDs of processes that need to be terminated on ctrl_c.
+//
+// This is necessary because typically, a shell receiving a ctrl_c (SIGINT) signal will terminate
+// the process and its children but not the processes spawned by them.
+//
+// In this testing tool, grandchild processes are required and need to be managed.
+// One approach is to use the process group ID (pgid) and the `killpg` syscall
+// to terminate all processes under the same pgid.
+//
+// The channel is used collect all pgids that need to be terminated upon the ctrl_c event;
+// otherwise, some processes could remain as zombie processes.
+//
+// The task will manage the receiver of the channel, while the sender will receive the handle of the task for reference.
+fn ctrl_c_handler() -> (tokio::task::JoinHandle<()>, mpsc::Sender<u32>) {
+    let (tx, mut rx) = mpsc::channel::<u32>(100);
+
+    let handle = tokio::task::spawn(async move {
+        tokio::signal::ctrl_c()
+            .await
+            .expect("Failed to listen for ctrl_c (SIGINT) signal");
+
+        while let Some(gpid) = rx.recv().await {
+            use nix::{sys::signal, unistd::Pid};
+            signal::killpg(Pid::from_raw(gpid as i32), Some(signal::Signal::SIGKILL))
+                .expect("Failed kill {gpid} process group id");
+        }
+    });
+    (handle, tx)
 }
```

Comment:
My gut feeling is that there is something wrong with this piece of code. I cannot verbalize it for now, but that doesn't matter too much, so I will let it pass here.

Reply:
mmmhhh, could you please be more explicit? Do you not prefer the channel used for passing the pgid around or having a floating future waiting for ctrl_c?

Reply:
Ok, I gave it a bit of thought. 

On some level, this is because you just return a channel. It's not immediatelly clear what is the contract of the function and what is the client supposed to do with the channel.

It's also hard to reason about what is happening. Linux is notoriously broken with respect to process management, and what's important here is that the process IDs can be reused. Given that, it's not clear what guarantees that you are killing the right process group. Esp. since you perform the kill in the drop and also here.

Also, the least important is the naming: you can read the name as if this is the handler, but it rather sets up the handler.

This is not super high stakes piece of code, and the edge case would be pretty rare, that's why I am ok to turn my blind eye on it.

Reply:
> It's not immediatelly clear what is the contract of the function and what is the client supposed to do with the channel

I will try to document the function better and provide a more descriptive name

> since you perform the kill in the drop and also here

That's absolutely true, things are being dropped twice. As you said, it could be a problem in rare cases. So, I decided to go for this simpler implementation to merge the pr quicker, as zombie processes are a really annoying problem. I will provide a more secure version in a follow-up PR

Reply:
I've understood better how `tokio::select!` deals with futures, so now no channel is required. Everything relies on the `drop()` method. Also, `wait_interrupt` now does only what is expected!